### DISCUSSION TITLE
GModelUtil.generateId - how does it work?

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#643] [General]
Hi,

this is not an important question, so don't take it to seriously, but it took me now more than an hour trying to understand the method `GModelUtil.generateId`  - and finally I give up.

The `GModelUtil` class yells at me:  _You never studied informatics? Than you're too stupid for this code!_

I see no sense in the param ID of the method

    generateId(final EClass eClass, final String id, final GModelState modelState) 
 
For my understanding the method simply counts the number of model classes of 'eClass' in the modelState and increases the number by 1. 

The method is used in the workflow example to generate a default label for new elements added to the model. The Id reflects the `elementTypeID` which is part of the class `CreateNodeOperationHandler`. The elemenetTypeID (for my understanding) defines the type of the Node. 

You could expect that the ElementTypeID is used to generate separate counters for each type. But only the eClass seems to be relevant. 
![image](https://user-images.githubusercontent.com/1339124/167224715-8c0a25f0-0bf2-46c5-945b-75874589a9f7.png)

I am now fine with this. But if someone could explain to me how the method works, then I'm happy. It's never to late to learn. And I promise to make a pull request with the corresponding  javadoc comment in the method header ;-)

**Comment by planger** [comment]
This is really nothing special, it just keeps an integer for each element type (e.g. an EClass) to provide a counter, so that unique names and IDs can be generated by creating id+counter, such as `ManualTask3`.
Essentially it is a map of `type to integer` that is kept in the index. So when a new instance for a type is created, you can create a new name for it by applying the lambda function `(i) -> id + i` (whereas `id` is the element type name and `i` is the next unused integer).
I admit it is not written very clearly.
